File: Data Structure/Binary Tree/Easy/LNT_no1106_Maximum Binary Tree.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param nums: an array
     * @return: the maximum tree
     */
    TreeNode * constructMaximumBinaryTree(vector<int> &nums) {
        // Write your code here
        return helper(nums, 0, nums.size() - 1);
    }

    TreeNode* helper(vector<int>& nums, int left, int right) {
        if (left > right)
            return nullptr;
        
        int index = left, value = nums[left];
        for (int i = left; i <= right; i++) {
            if (nums[i] > value) {
                value = nums[i];
                index = i;
            }
        }

        TreeNode* node = new TreeNode(value);
        node->left = helper(nums, left, index - 1);
        node->right = helper(nums, index + 1, right);

        return node;
    }
};

File: Data Structure/Binary Tree/Easy/LNT_no1106_Maximum Binary Tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param nums: an array
    @return: the maximum tree
    """
    def constructMaximumBinaryTree(self, nums):
        # Write your code here
        return self.helper(nums, 0, len(nums) - 1)
    
    def helper(self, nums, left, right):
        if left > right:
            return None
        
        index, value = left, nums[left]
        for i in range(left, right + 1):
            if nums[i] > value:
                index, value = i, nums[i]
        
        node = TreeNode(value)
        node.left = self.helper(nums, left, index - 1)
        node.right = self.helper(nums, index + 1, right)

        return node
